src/Integration/LocalServices/ProjectSystemHelper.cs

```diff
@@ -502,5 +492,47 @@ public bool IsLegacyProjectSystem(Project dteProject)
 
             return hierarchy != null && hierarchy is IVsAggregatableProjectCorrected;
         }
+
+        public IEnumerable<VSConstants.VSITEMID> GetAllItems(IVsHierarchy vsHierarchy)
+        {
+            // Based on https://github.com/microsoft/VSSDK-Extensibility-Samples/blob/c6b80a4ff7023578649d7edecc8fd6cd8a34da10/Code_Sweep/C%23/VsPackage/BuildManager.cs#L450
+            return ChildrenOf(vsHierarchy, (uint) VSConstants.VSITEMID.Root);
+        }
+
+        private static IEnumerable<VSConstants.VSITEMID> ChildrenOf(IVsHierarchy hierarchy, uint rootId)
+        {
+            var result = new List<VSConstants.VSITEMID>();
+
+            for (var itemId = FirstChild(hierarchy, rootId);
+                itemId != (uint)VSConstants.VSITEMID.Nil;
+                itemId = NextSibling(hierarchy, itemId))
+            {
+                result.Add((VSConstants.VSITEMID) itemId);
+                result.AddRange(ChildrenOf(hierarchy, itemId));
+            }
+
+            return result;
+        }
+
+        private static uint FirstChild(IVsHierarchy hierarchy, uint rootId) =>
+            GetItemId(hierarchy, rootId, __VSHPROPID.VSHPROPID_FirstChild);
+
+        private static uint NextSibling(IVsHierarchy hierarchy, uint firstId) =>
+            GetItemId(hierarchy, firstId, __VSHPROPID.VSHPROPID_NextSibling);
+
+        private static uint GetItemId(IVsHierarchy hierarchy, uint itemId, __VSHPROPID propId)
+        {
+            hierarchy.GetProperty(itemId, (int)propId, out var idObj);
+
+            return idObj == null ? (uint)VSConstants.VSITEMID.Nil : (uint)(int) idObj;
+        }
+
+        public string GetItemFilePath(IVsProject vsProject, VSConstants.VSITEMID itemId)
+        {
+            var hr = vsProject.GetMkDocument((uint)itemId, out var filePath);
```

Comment:
@duncanp-sonar I did some testing in a c++ project and I get a debug exception here.
It seems to be that in c++ projects, `GetAllItems` returns also all included files, such as framework `C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\shared\basetsd.h`. I get hundreds of files in my tiny project, whilst with DTE I get only the relevant ones. I'm not sure how to fix `GetAllItems` to ignore all the c++ dependencies. I'll look into it and won't merge this PR for now.